refactor: Remove redundant and confusing calls to IsArgSet

maflcko opened the pull request:
`IsArgSet` is problematic:

* It returns whether an arg has been set, even if it has been negated. `IsArgSet` is sometimes used to check for a truthy value, which is wrong, but usually harmless. Cleanup of those cases may or may not be done in a follow-up.
* In most other cases, calling it is redundant, because the immediately following `Get*Arg` calls can already return an `std::optional` nullopt value to indicate an unset arg.

So relieve both issues by removing all `IsArgSet` that are redundant.

Commit:
refactor: Remove IsArgSet guard when fallback value is provided

Checking for IsArgSet before calling GetArg while providing the args
default value as fallback is both confusing and fragile.

It is confusing, because the provided fallback is dead code. So it would
be better to just call GetArg without a fallback.

However, ignoring the fallback value is fragile, because it would not be
sanitized.

Fix all issues by sanitizing the fallback value.

Commit:
refactor: Remove redundant call to IsArgSet

Checking for IsArgSet before calling GetArg while providing an arbitrary
default value as fallback is both confusing and fragile.

It is confusing, because the provided fallback is dead code. So it would
be better to just call GetArg without a fallback.

Even better would be to provide the true fallback value and sanitize it
as if it were user-input, but this can be done in a follow-up.

Removing the redundant call to IsArgSet will have to be done either way,
so do it now.

ryanofsky reviewed (approved):
Code review ACK fa3fb3c23fae287b161112b9b04cf0937a1051c6. Nice refactoring to get rid of pointless complexity and repetition in this code.

yancyribbens commented:
> It is confusing, because the provided fallback is dead code. So it would be better to just call GetArg without a fallback.

I'm not 100% sure when reading the commit what is meant by dead code.  Is it that the arg `-blockmintxfee` is a no longer used switch here?  I ask because I notice some other calls to GetArg with fallbacks such as:

```
`src/init.cpp:879:    if (args.IsArgSet("-upnp")) {`
``

yancyribbens commented on src/init.cpp:
```diff
@@ -1012,20 +1012,20 @@ bool AppInitParameterInteraction(const ArgsManager& args)
 
     // Sanity check argument for min fee for including tx in block
     // TODO: Harmonize which arguments need sanity checking and where that happens
```
Is this TODO still accurate?  Not what arguments need Harmonizing.

yancyribbens replied:
I agree the TODO isn't very clear and possibly no longer applicable.  Maybe it could be rounded up with other unclear comments and TODOs in a separate PR.  A PR to just remove this one TODO line seems silly.

ryanofsky commented:
> > It is confusing, because the provided fallback is dead code. So it would be better to just call GetArg without a fallback.
> 
> I'm not 100% sure when reading the commit what is meant by dead code.

It is just saying the fallback argument is unused. For example in:

```c++
   if (gArgs.IsArgSet("-color")) {
   {
        const std::string color{gArgs.GetArg("-color", DEFAULT_COLOR_SETTING)};
        ...
   }
```

The `DEFAULT_COLOR_SETTING` argument is unused, and any fallback value that is passed would be ignored because `IsArgSet` returned true.

yancyribbens commented:
Concept ACK https://github.com/bitcoin/bitcoin/pull/31896/commits/fa3fb3c23fae287b161112b9b04cf0937a1051c6

jonatack commented:
Concept ACK

pablomartin4btc reviewed (commented):
cr & tACK fa3fb3c23fae287b161112b9b04cf0937a1051c6

What about the `IsArgSet` ones followed by `GetPathArg`?

Commit:
doc: Remove outdated and stale todo comment

If anything is left to be done, a new discussion issue or pull request
can be created.

pablomartin4btc reviewed (approved):
re-ACK fa0cf040baa74b161b4b00ddea95bfb5aac0bb53

Since my last review: outdated TODO [comment](https://github.com/bitcoin/bitcoin/pull/31896#discussion_r1997633914) was removed (on new commit) - no impact.

fanquake commented:
@ryanofsky Want to take another look?

ryanofsky reviewed (approved):
Code review ACK fa0cf040baa74b161b4b00ddea95bfb5aac0bb53. Only change since last review is removing a nearby todo comment that was vague

jonatack commented on src/init.cpp:
```diff
@@ -1157,11 +1157,10 @@ bool CheckHostPortOptions(const ArgsManager& args) {
     }) {
-        if (args.IsArgSet(port_option)) {
-            const std::string port = args.GetArg(port_option, "");
+        if (const auto port{args.GetArg(port_option)}) {
```
For all `GetArg` calls in this commit. Perhaps I am confused but should be const ref? Would also be more clear to continue to specify the type here.

```suggestion
        if (const std::string& port{args.GetArg(port_option)}) {
```

jonatack commented on src/init.cpp:
```diff
@@ -1010,8 +1010,6 @@ bool AppInitParameterInteraction(const ArgsManager& args)
     }
 
-    // Sanity check argument for min fee for including tx in block
-    // TODO: Harmonize which arguments need sanity checking and where that happens
```
FWIW these two lines are from daec955fd68b (https://github.com/bitcoin/bitcoin/pull/9380)

jonatack reviewed (commented):
Concept ACK.

Part of the commit message of fa41685f438bba00761423d464bbb8dc5ea7ddf6 seems incorrect (maybe was pasted from fa3fb3c23fae287b161112b9b04cf0937a1051c6 without updating, not sure).

pablomartin4btc reviewed (approved):
re-ACK 0000fb3fd9f01f8a8cbe9de0bdc080023b6f00fc

No changes, just rebased (basically due to this #31278).

ryanofsky reviewed (approved):
Code review ACK 0000fb3fd9f01f8a8cbe9de0bdc080023b6f00fc. No changes since last review other than rebase.
